app/src/commonMain/kotlin/io/github/cgr/app/Assembly.kt

```diff
@@ -0,0 +1,105 @@
+package io.github.cgr.app
+
+import io.github.cgr.app.io.MAGIC_NUMBER
+import io.github.cgr.app.io.Packer
+import io.github.cgr.app.objects.BaseLabel
+import io.github.cgr.app.objects.Opcode
+
+fun assemble(program: String): ByteArray {
+    registerBase()
+    val opcodes = program.lines()
+        .asSequence()
+        .map(String::trim)
+        .filter(String::isNotEmpty)
+        .filter { !it.startsWith('#') }
+        .map(::parseLine)
+        .toList()
+    val constants = opcodes.flatMap { it.second }
+        .filter { (it.startsWith('"') && it.endsWith('"')) || (it.startsWith('[') && it.endsWith(']')) }
+        .distinct()
+    val constantMap = constants.mapIndexed { i, s -> s to i }.toMap()
+    val (builtPool, labels) = buildConstantPool(constants.map { it.substring(1, it.length - 1) })
+    val symbolTable = (labels + opcodes.map { it.first }).distinct()
+    val symbolMap = symbolTable.mapIndexed { i, s -> s to i }.toMap()
```

Comment:
I can't help but think that a BidirectionalMap type would make this significantly easier. I don't think the JVM lets you say `class BidirectionalMap<T, U> implements Map<T, U>, Map<U, T>` though.

Reply:
Really, the maps aren't needed, I could just use `indexOf`. Implementing a full on new map type just for this would be wasteful anyway.